### DISCUSSION TITLE
Hovernet gives strange contours coordinates

### DISCUSSION THREAD

**Discussion by OmarAshkar** [Project-MONAI/MONAI#6466] [Q&A]
Hello,

I am having a Hovernet post-processing workflow that I try to optimize it alittle. I had to cut down the images into 4000x4000 images before running the inference. After that, I have exported the coords contours that comes from  `HoVerNetInstanceMapPostProcessingd`. Now the numbers doesn't make any sense to me as I said the image were 4000, but he contours coords get to be more than 20,000! What maybe wrong here? 

Another thing that the contours are not "succinate"  as I can see from the implementation, so I don't know how to convert them back again to tamed contours.

Also, can I just turn off the transformers and save the two maps so I can do the statistics myself without monai? When I cannot remove `HoVerNetInstanceMapPostProcessingd` from the transformers the instance map is not exported.

Any help is appreciated. 

Thanks in advance. 

Omar

# EDIT 

I have replace the `HoVerNetInstanceMapPostProcessingd` with the following code, but I get error for duplicate keys.
```
Compose([
            GenerateWatershedMaskd(keys = ["image"]),
            GenerateInstanceBorderd(hover_map_key= HoVerNetBranch.NP.value),
            GenerateDistanceMapd(),
            GenerateWatershedMarkersd(radius = 1),
            Watershedd(keys = ["dist_map"], markers_key= "markers"),
            HoVerNetNuclearTypePostProcessingd()
])

```
I am not sure if I wrote that correctly. Unfortunately the keys argument is always very generic.

It fails on the HoVerNetNuclearTypePostProcessingd() as the inst_info key was not generated. Also, I want to save the type_map and instance_map. I assume the instance_map will be dist_map here, but how about the updated type_map to save?

**Comment by OmarAshkar** [comment]
Hi 

I have the following update on my post_transformers. Despite that this runs fine, the instance_map and type_map does not seem right to me. 

```
    post_transforms = Compose(
        [
            FlattenSubKeysd(
                keys="pred",
                sub_keys=[HoVerNetBranch.NC.value, HoVerNetBranch.NP.value, HoVerNetBranch.HV.value],
                delete_keys=True,
            ),
            GenerateWatershedMaskd(min_object_size = min_object_size, 
                                   threshold = mask_threshold,
                                    keys = [HoVerNetBranch.NP.value]), 
            GenerateInstanceBorderd(kernel_size = sobel_kernel_size, hover_map_key= HoVerNetBranch.HV.value),
            GenerateDistanceMapd(), 
            GenerateWatershedMarkersd(min_object_size = min_object_size, radius = marker_radius, threshold = marker_threshold),
            Watershedd(keys = ["dist_map"], markers_key= "markers", mask_key= "mask"),
            Activationsd( keys=HoVerNetBranch.NC.value, softmax  = True),
            AsDiscreted(keys=HoVerNetBranch.NC.value, argmax  = True),
            FromMetaTensord(keys=["image", "type_prediction"]),
            SaveImaged(
                keys = "dist_map",
                meta_keys="image_meta_dict",
                output_ext=".nii.gz",
                output_dir=output_dir,
                output_postfix="instance_map",
                output_dtype="uint32",
                separate_folder=False
            ),
            SaveImaged(
                keys = "type_prediction",
                meta_keys="image_meta_dict",
                output_ext=".nii.gz",
                output_dir=output_dir,
                output_postfix="type_map",
                output_dtype="uint8",
                separate_folder=False
            ),
        ]
    )
```

I just want to remove the instance_info calculations as it is fairly slow. Related to #6046 .

Thanks

**Reply by KumoLiu** [reply]
Hi @OmarAshkar, one suggestion was to generate each of these intermediate outputs, then adjust the hyperparameters(such as sobel_kernel_size, sobel_kernel_size, mask_threshold, and so on) since the hyperparameters setting in this post-processing is very task-specific.
https://github.com/Project-MONAI/MONAI/blob/e0d0ce7d94fc63f3d32fdf03520048f6ae4acaab/monai/apps/pathology/transforms/post/array.py#L722-L726

Hope it can help you, thanks!

**Reply by OmarAshkar** [reply]
Hi @KumoLiu. I just used the same values. But HoVerNetInstanceMapPostProcessingd gives total different (more realistic) results. I just want to turn off the measurments steps as they take a lot of time. Do you see here if I am coding something wrong other than the parameters?

**Reply by KumoLiu** [reply]
Hi @OmarAshkar, from the code I didn't see any obvious wrong. If could just remove these post-processing transforms and just use some of them. Just make sure that you have noticed that from the HoVerNet, it outputs three branches containing binary segmentation, pixel-level classification map, and hovermap.  Thanks!
